src/peft/tuners/oft/layer.py

```diff
@@ -431,42 +637,15 @@ def forward(self, x: torch.Tensor, *args, **kwargs) -> torch.Tensor:
         elif self.merged:
             result = self.base_layer(x, *args, **kwargs)
         else:
-            oft_rotation = torch.eye(self.in_features, device=x.device)
-            oft_scale = torch.ones((int(self.out_features), 1), device=x.device)
-
             for active_adapter in self.active_adapters:
-                if active_adapter not in self.oft_r.keys():
+                if active_adapter not in self.oft_R.keys():
                     continue
-                oft_r = self.oft_r[active_adapter]
-                oft_s = self.oft_s[active_adapter]
-                dropout = self.oft_dropout[active_adapter]
+                oft_R = self.oft_R[active_adapter]
 
-                rank = self.r[active_adapter]
-                coft = self.coft[active_adapter]
-                eps = self.eps[active_adapter]
+                x = self._cast_input_dtype(x, oft_R.weight.dtype)
+                x = oft_R(x)
 
-                if coft:
-                    with torch.no_grad():
-                        oft_r.copy_(self._project_batch(oft_r, eps=eps))
-
-                orth_rotate = self._cayley_batch(oft_r)
-                orth_rotate = dropout(orth_rotate)
-                oft_mat = self._block_diagonal(orth_rotate, rank)
-
-                oft_rotation = oft_mat @ oft_rotation
-                oft_scale = oft_s * oft_scale
-
-            x = x.to(self.get_base_layer().weight.data.dtype)
-
-            orig_weight = self.get_base_layer().weight.data
-            orig_weight = torch.transpose(orig_weight, 0, 1)
-            rotated_weight = torch.mm(oft_rotation, orig_weight.to(oft_rotation.dtype))
-            rotated_weight = torch.transpose(rotated_weight, 0, 1)
-            scaled_rotated_weight = rotated_weight * oft_scale
-
-            x = self._cast_input_dtype(x, scaled_rotated_weight.dtype)
-            bias = self._cast_input_dtype(self.get_base_layer().bias, scaled_rotated_weight.dtype)
-            result = F.linear(input=x, weight=scaled_rotated_weight, bias=bias)
+            result = self.base_layer(x, *args, **kwargs)
```

Comment:
So this line errors for me when the base layer is bfloat16. The reason is that x has been upcast to float32 due to `x = self._cast_input_dtype(x, oft_R.weight.dtype)`. I could fix the problem by changing this line to:

```suggestion
            result = self.base_layer(x.to(previous_dtype), *args, **kwargs)
```

Reply:
Thanks, updated.